# PR Testing Report

## PR Information

- **PR Number:** #18991
- **Title:** Remove misleading TerminalOptions.Shell no-op
- **Head Commit:** `2430a7f76c422321059c3fb5a8f8dd5fd81412a4`
- **Tested At:** 2026-08-04T13:09:45Z

## Artifact Version Verification

- **Expected Commit:** `2430a7f76c422321059c3fb5a8f8dd5fd81412a4`
- **Installed Version:** `13.5.0-pr.18991.g2430a7f7`
- **Status:** ✅ Verified

The CLI and PR NuGet hive were installed into an isolated local temp directory with the PR dogfood installer. The installed version contains the PR head's short SHA.

## Changes Analyzed

### Files Changed

- `src/Aspire.Hosting/ApplicationModel/TerminalAnnotation.cs`
- `src/Aspire.Hosting/TerminalResourceBuilderExtensions.cs`
- `src/Aspire.TerminalHost/TerminalHostApp.cs`
- `src/Aspire.TerminalHost/TerminalHostArgs.cs`
- `tests/Aspire.Hosting.Tests/WithTerminalTests.cs`
- `tests/Aspire.TerminalHost.Tests/TerminalHostArgsTests.cs`

### Change Categories

- [ ] CLI changes
- [x] Hosting API changes
- [ ] Dashboard changes
- [ ] Template changes
- [ ] Client/component changes
- [ ] VS Code extension changes
- [x] Test changes
- [ ] CI infrastructure changes

## Test Scenarios Executed

### Scenario 1: Executable terminal happy path

**Objective:** Verify that `WithTerminal()` attaches to the executable resource's own process and preserves supported terminal dimensions after removing shell selection.

**Coverage Type:** Happy path

**Status:** ✅ Passed

**Steps:**

1. Created a fresh C# `aspire-empty` AppHost from the PR hive.
2. Added `/bin/sh` directly with `AddExecutable(...).WithTerminal(...)`.
3. Set `Columns = 100` and `Rows = 30`.
4. Started the AppHost with the PR CLI and waited for `shell` to reach `up`.
5. Attached with `aspire terminal attach shell`.
6. Waited for primary terminal control, sent a marker command through the PTY, and verified `EXEC_TERMINAL_OK`.
7. Exited the resource and stopped the AppHost.

**Evidence:**

- `scenario-executable/start-output.txt`
- `scenario-executable/wait-output.txt`
- `scenario-executable/describe-output.txt`
- `scenario-executable/terminal-expect-output.txt`
- `scenario-executable/terminal-expect.log`
- `scenario-executable/stop-output.txt`

**Observations:**

- The executable resource reached `Running` / `Healthy`.
- The attach client connected to replica 0 and acquired primary control.
- `/bin/sh` executed the marker command and returned the expected output.
- This confirms that the executable itself is the terminal program; no separate shell selection is involved.

---

### Scenario 2: Container terminal happy path

**Objective:** Verify the corresponding container behavior and real DCP container terminal spec path.

**Coverage Type:** Happy path

**Status:** ✅ Passed

**Steps:**

1. Created a separate fresh C# `aspire-empty` AppHost from the PR hive.
2. Added `alpine:3.22` with `sh` as its entrypoint and enabled `WithTerminal(...)`.
3. Set `Columns = 100` and `Rows = 30`.
4. Started the AppHost and waited for the container resource to reach `up`.
5. Attached with `aspire terminal attach shell`.
6. Waited for primary terminal control, sent paced input through the PTY, and verified `CONTAINER_TERMINAL_OK`.
7. Exited the container shell and stopped the AppHost.

**Evidence:**

- `scenario-container/start-output.txt`
- `scenario-container/wait-output.txt`
- `scenario-container/describe-output.txt`
- `scenario-container/terminal-expect-base64-output.txt`
- `scenario-container/terminal-expect-base64.log`
- `scenario-container/stop-output.txt`

**Observations:**

- The container resource reached `Running` / `Healthy`.
- Alpine's `/ #` prompt was visible through the terminal attach session.
- The shell accepted PTY input and produced the expected marker output.

---

### Scenario 3: Removed Shell API

**Objective:** Verify that the experimental API contract no longer exposes a misleading shell-selection property.

**Coverage Type:** Unhappy path

**Status:** ✅ Passed

**Steps:**

1. Created a third fresh C# `aspire-empty` AppHost from the PR hive.
2. Configured `WithTerminal(options => options.Shell = "/bin/bash")`.
3. Built the AppHost and required a non-zero exit code.
4. Verified the compiler reported `CS1061` specifically for missing `TerminalOptions.Shell`.

**Expected Unhappy-Path Outcome:** A clear compile-time error directing consumers away from the removed no-op API.

**Evidence:**

- `scenario-removed-api/build-output.txt`

**Observed Error:**

```text
error CS1061: 'TerminalOptions' does not contain a definition for 'Shell'
```

## Additional Validation

- Repository-focused tests previously run on the PR branch:
  - `WithTerminalTests`: 23 passed
  - Executable and container DCP terminal-spec tests: 2 passed
  - `TerminalHostArgsTests`: 12 passed
- GitHub Actions at report time:
  - 337 passed
  - 3 skipped
  - 0 pending
  - 0 failed
- No test AppHosts remained running after cleanup.
- The repository worktree remained clean.

## Environment Notes

The installed Hex1b CLI could not start because `libhex1binterop.dylib` is unavailable on this host. The scenarios used macOS `/usr/bin/expect` instead, which still allocated a real PTY, synchronized on the attach client's primary-control state, sent terminal input, and captured output.

## Summary

| Scenario | Status | Notes |
|----------|--------|-------|
| Executable terminal happy path | ✅ Passed | Resource process handled PTY input and emitted marker |
| Container terminal happy path | ✅ Passed | Alpine `sh` handled PTY input and emitted marker |
| Removed Shell API | ✅ Passed | Expected `CS1061` for `TerminalOptions.Shell` |

## Overall Result

**✅ PR VERIFIED**

The PR artifact matches the latest commit. Both executable and container terminal sessions work with the supported model/spec behavior, and the removed experimental shell-selection API fails clearly at compile time.

--- a/src/Aspire.TerminalHost/TerminalHostArgs.cs
+++ b/src/Aspire.TerminalHost/TerminalHostArgs.cs
@@ -41,6 +41,7 @@ internal sealed class TerminalHostArgs
     ///   <item><c>--control-uds PATH</c> (required) — path the host LISTENS on; AppHost dials for status/shutdown RPC.</item>
     ///   <item><c>--columns N</c> (optional, default 120)</item>
     ///   <item><c>--rows N</c> (optional, default 30)</item>
+    ///   <item><c>--shell NAME</c> (optional, accepted for compatibility and ignored)</item>
     /// </list>
     /// Every option is single-valued and may only be specified once; duplicates throw
     /// <see cref="TerminalHostArgsException"/>. We use <c>System.CommandLine</c> so the
@@ -62,6 +63,11 @@ internal sealed class TerminalHostArgs
             "Initial PTY width in columns (default 120).", defaultValue: 120);
         var rowsOption = SingleValueOption<int>("--rows", required: false,
             "Initial PTY height in rows (default 30).", defaultValue: 30);
+        // Older Aspire.Hosting packages can run with a newer CLI-provided terminal host
+        // and still emit --shell. Accept the argument so that mixed-version AppHosts start,
+        // but ignore it because DCP launches the resource process that owns the PTY.
+        var shellOption = SingleValueOption<string?>("--shell", required: false,
+            "Legacy shell hint accepted for compatibility and ignored.");
 
         // Cols / rows must be positive. System.CommandLine has no built-in range validator
         // for ints, so attach one explicitly. Skip when the value couldn't be converted at
@@ -98,6 +104,7 @@ internal sealed class TerminalHostArgs
             controlOption,
             columnsOption,
             rowsOption,
+            shellOption,
         };
         // The terminal host argv comes from DCP only; treat unknown flags as a hard error
         // so we don't silently accept garbage and start with the wrong configuration.
@@ -167,4 +174,3 @@ internal sealed class TerminalHostArgs
 /// Thrown when the terminal host receives malformed command-line arguments.
 /// </summary>
 internal sealed class TerminalHostArgsException(string message) : Exception(message);
-

--- a/tests/Aspire.TerminalHost.Tests/TerminalHostArgsTests.cs
+++ b/tests/Aspire.TerminalHost.Tests/TerminalHostArgsTests.cs
@@ -37,6 +37,21 @@ public class TerminalHostArgsTests
     }
 
     [Fact]
+    public void ParseAcceptsIgnoredShellForCompatibility()
+    {
+        var args = TerminalHostArgs.Parse([
+            "--producer-uds", "/tmp/p.sock",
+            "--consumer-uds", "/tmp/c.sock",
+            "--control-uds", "/tmp/ctrl.sock",
+            "--shell", "/bin/bash",
+        ]);
+
+        Assert.Equal("/tmp/p.sock", args.ProducerUdsPath);
+        Assert.Equal("/tmp/c.sock", args.ConsumerUdsPath);
+        Assert.Equal("/tmp/ctrl.sock", args.ControlUdsPath);
+    }
+
+    [Fact]
     public void ParseMissingProducerUdsThrows()
     {
         var ex = Assert.Throws<TerminalHostArgsException>(() => TerminalHostArgs.Parse([

